chain-impl-mockchain/src/txbuilder.rs

```diff
@@ -0,0 +1,192 @@
+        self.0.inputs.push(input.clone())
+    }
+
+    /// Add additional output.
+    ///
+    /// Each output may extend the size of the required fee.
+    pub fn add_output(&mut self, address: Address, value: Value) {
+        self.0.outputs.push(tx::Output { address, value })
+    }
+
+    /// We finalize the transaction by passing fee rule and return
+    /// policy. Then after all calculations were made we can get
+    /// the information back to us.
+    ///
+    pub fn finalize<F: FeeAlgorithm>(
+        mut self,
+        fee_algorithm: F,
+        policy: OutputPolicy,
+    ) -> Result<(Balance, TransactionFinalizer), Error> {
+        // calculate initial fee, maybe we can fit it without any
+        // additional calculations.
+        let fee = fee_algorithm.calculate_for(&self.0);
+        let pos = match balance(&self.0, fee) {
+            Ok(Balance::Negative(_)) => return Err(Error::TxNotEnoughTotalInput),
+            Ok(Balance::Positive(v)) => v,
+            Ok(Balance::Zero) => {
+                return Ok((Balance::Zero, TransactionFinalizer::new(self.0)));
+            }
+            Err(err) => return Err(Error::MathErr(err)),
+        };
+        // we have more money in the inputs then fee and outputs
+        // so we need to return some money back to us.
+        match policy {
+            OutputPolicy::Forget => {
+                let tx = TransactionFinalizer(tx::SignedTransaction {
+                    transaction: self.0,
+                    witnesses: vec![],
+                });
+                Ok((Balance::Positive(pos), tx))
+            }
+            // We will try to find the best matching value, for
+            // this reason we will try to reduce the set using
+            // value estimated by the current fee.
+            //
+            // We are searching in a range
+            //   [min_value, max_value)
+            OutputPolicy::One(address) => {
+                // This is simplified version of the algorithm that
+                // works only in case in fee can perfectly estimate
+                // the required cost. We add an additional empty output
+                // hoping that it doesn't change fee count.
+                //
+                // Otherwise better estimation algorithm is needed.
+                let mut tx = self.0.clone();
+                tx.outputs.push(tx::Output {
+                    address: address.clone(),
+                    value: Value(0),
+                });
+                let fee = fee_algorithm.calculate_for(&tx);
+                match balance(&tx, fee) {
+                    Ok(Balance::Positive(value)) => {
+                        self.0.outputs.push(tx::Output { address, value });
+                        return Ok((Balance::Zero, TransactionFinalizer::new(self.0)));
+                    }
+                    _ => {
+                        return Ok((Balance::Positive(pos), TransactionFinalizer::new(tx)));
```

Comment:
When balance with 0-value TX turns negative, do we still want to perform the 0-value TX? It will get rejected. This strategy seems fine when the result is `Balance::Zero`. I'm not sure what to do if balance is negative. Remove 0-value TX and add the excessive value to a random output? Assign excess to output address 0 and send it to junkyard forever littering UTXO space?

Reply:
The situation here, is that few lines above we have already generated transaction with the positive value, so we can always fall back to that transaction.

So I had the following consideration: creating empty transaction doesn't benefit the sender at all so in `Zero` case we fall back to the previous approach, where we had a bit of positive balance. This can potentially be interesting to the node performing transaction.

In general, if the fee calculation procedure is more complex we will need to perform clever branch analysis.

Reply:
But as far as I can see in this specific case (line 125) we return `tx`, which has zero value output added (line 114), we don't fall back to the slightly positive transaction `self.0` without zero value output.

Reply:
Thanks, it should be `self.0` here!

Reply:
What happens with the excess if we create transaction with positive balance? Does it get lost forever? Are zero-value outputs ultimate dust or they get discarded before reaching UTXO and therefore safe to generate?

Reply:
As @NicolasDP it's perfectly ok and we can loose money, also as far as I know that can be used by the node applying a transaction so excess of the money may be  used to increase transaction priority as nodes may be more interested to commit that one.